src/clp_ffi_js/ir/utils.hpp

```diff
@@ -0,0 +1,165 @@
+#define CLP_FFI_JS_IR_UTILS_HPP
+
+#include <algorithm>
+#include <cstddef>
+#include <string>
+#include <string_view>
+#include <type_traits>
+#include <vector>
+
+#include <clp/ErrorCode.hpp>
+#include <clp/TimestampPattern.hpp>
+#include <clp/TraceableException.hpp>
+#include <clp/type_utils.hpp>
+#include <emscripten/em_asm.h>
+#include <emscripten/val.h>
+
+#include <clp_ffi_js/ClpFfiJsException.hpp>
+#include <clp_ffi_js/constants.hpp>
+#include <clp_ffi_js/ir/StreamReader.hpp>
+#include <clp_ffi_js/ir/StructuredIrStreamReader.hpp>
+#include <clp_ffi_js/ir/UnstructuredIrStreamReader.hpp>
+
+namespace clp_ffi_js::ir {
+constexpr std::string_view cEmptyJsonStr{"{}"};
+
+/**
+ * A templated function that implements `StreamReader::filter_log_events` for both
+ * `UnstructuredIrStreamReader` and `StructuredIrStreamReader`. The additional arguments,
+ * which are not part of `filter_log_events`, are private members of the derived
+ * `StreamReader` classes.
+ *
+ * @param log_level_filter
+ * @param log_events
+ * @param[out] filtered_log_event_map A reference to `FilteredLogEventsMap` that stores filtered
+ * result.
+ */
+template <typename LogEvents>
+auto generic_filter_log_events(
+        FilteredLogEventsMap& filtered_log_event_map,
+        LogLevelFilterTsType const& log_level_filter,
+        LogEvents const& log_events
+) -> void {
+    if (log_level_filter.isNull()) {
+        filtered_log_event_map.reset();
+        return;
+    }
+
+    filtered_log_event_map.emplace();
+    auto filter_levels
+            = emscripten::vecFromJSArray<std::underlying_type_t<LogLevel>>(log_level_filter);
+
+    for (size_t log_event_idx = 0; log_event_idx < log_events.size(); ++log_event_idx) {
+        auto const& log_event = log_events[log_event_idx];
+        if (std::ranges::find(
+                    filter_levels,
+                    clp::enum_to_underlying_type(log_event.get_log_level())
+            )
+            != filter_levels.end())
+        {
+            filtered_log_event_map->emplace_back(log_event_idx);
+        }
+    }
+}
+
+/**
+ * A templated function that implements `StreamReader::decode_range` for both
+ * `UnstructuredIrStreamReader` and `StructuredIrStreamReader`. The additional
+ * arguments, which are not part of `decode_range`, are private members of
+ * the derived `StreamReader` classes.
+ *
+ * @param begin_idx
+ * @param end_idx
+ * @param filtered_log_event_map
+ * @param log_events
+ * @param use_filter
+ * @param ts_pattern A pattern for formatting unstructured log event timestamps as date strings.
+ * The pattern is unused for structured log events, as they are not formatted in `clp-ffi-js`.
+ * @return
+ */
+template <typename LogEvents>
+auto generic_decode_range(
+        size_t begin_idx,
+        size_t end_idx,
+        FilteredLogEventsMap const& filtered_log_event_map,
+        LogEvents const& log_events,
+        bool use_filter,
+        clp::TimestampPattern const& ts_pattern
```

Comment:
@kirkrodrigues 
1. Shall we use `std::enable_if<std::is_same_v<>>` for `ts_pattern`?
2. if not, do we consider Pack / variadic arguments a bad coding practice in C++? can that be used here?
3. if not, can we create a generic DecoderOptions class?

Reply:
1. Sure
2. clang-tidy [discourages](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/pro-type-vararg.html) variadic functions.
3. Sure, but how would that help?

Reply:
1. I was thinking we can declare separate function signatures (one that is specialized for class UnstructuredStreamReader and the other one(s) is/are for class StructuredStreamReader and others. For the common logics, both of them call the same helpers.
2. In a way I see how (1.) can be considered a bad practice for similar reasons. It could be hard to maintain when we add more stream reader classes.
3. If I don't picture things wrong, in the StreamReader-type templated DecoderOptions class we can enable attributes only for certain StreamReader types. Something like
    ```cpp
    template<typename StreamReader>
    class DecoderOptions {
    public:
        // Other members can be added here
    };
    
    // Specialization for UnstructuredStreamReader
    template<>
    class DecoderOptions<UnstructuredStreamReader> {
    public:
        clp::TimestampPattern const& ts_pattern;
    
        DecoderOptions(const clp::TimestampPattern& ts_pattern_ref) 
            : ts_pattern(ts_pattern_ref) {}
    };
    ```
    
    i guess we can also do
    ```
    template<typename StreamReader>
    class DecoderOptions {
    public:
        // Constructor specialization for UnstructuredStreamReader
        template<typename T = StreamReader, typename = std::enable_if_t<std::is_same_v<T, UnstructuredStreamReader>>>
        DecoderOptions(const clp::TimestampPattern& ts_pattern_ref) 
            : ts_pattern(&ts_pattern_ref) {}
    
        // or make this std::Optional
        const clp::TimestampPattern* ts_pattern = nullptr;
    };
    ```

Reply:
Let's continue the discussion at https://github.com/y-scope/clp-ffi-js/pull/35/files#r1882227286 before we look at the approaches proposed in this thread.